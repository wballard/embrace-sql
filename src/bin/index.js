#!/usr/bin/env node

// bootstrap typescript from the command line without a prebuild
require("ts-node").register({
  /* options */
});
// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require("path");

require(path.join(__dirname, "..", "cli", "index"));
