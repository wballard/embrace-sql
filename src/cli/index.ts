import { program } from "commander";
import pj from "../../package.json";
import init from "./init";
import start from "./start";
import migrate from "./migrate";
import { install } from "../structured-console";

/**
 * Our little command line interface. This can be used from Docker -- which is the
 * intended use, or with `yarn cli` for development and testing.
 */
const main = async (): Promise<void> => {
  install();
  program.version(pj.version);
  program.description(`Command line interface to EmbraceSQL.`);
  program.addCommand(init);
  program.addCommand(start);
  program.addCommand(migrate);
  // GO!
  await program.parseAsync(process.argv);
};
main();
