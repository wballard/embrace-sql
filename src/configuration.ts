import { cosmiconfig } from "cosmiconfig";
import Url from "url-parse";
import process from "process";
import path from "path";
const debug = require("debug")("embracesql:configuration");
import embraceDatabases, { DatabaseInstance } from "./database-engines";

/**
 * The all important root configuration. This tells EmbraceSQL which databases to manage.
 */
export type Configuration = {
  /**
   * The root directory used to start this config.
   */
  embraceSQLRoot: string;
  /**
   * URL identifying Kafka topic or REST endpoing to post commands.
   */
  twinCommandsTo?: string;
  /**
   * All available databases.
   */
  databases: Map<string, Url>;
};

/**
 * This is the default to set up a new context on each API invocation, as well as 'the' context
 * for internal code generation.
 *
 * The type here is a bit different as the context used in 'client code' is generated
 * with database type names. Here on the inside
 */
export type RootContext = {
  /**
   * All configured databases, by name. This is the internal root context, so this is a hash and
   * not named properties. Client contexts will be generated with names to provide awesome autocomplete.
   */
  databases: Map<string, DatabaseInstance>;
};

/**
 * Load up a configuration object.
 *
 * This will interpolate environment variables, build and object and hand it back.
 *
 * Will look in the current directory or a environment variable set root.
 */
export const loadConfiguration = async (): Promise<Configuration> => {
  const root = path.normalize(process.env.EMBRACESQL_ROOT || process.cwd());
  debug(root);
  // TODO env var substition loader hook
  const explorer = cosmiconfig("embracesql", {
    searchPlaces: ["embracesql.yaml", "embracesql.yml"],
  });
  const result = await explorer.search(root);
  const config = result.config as Configuration;
  config.embraceSQLRoot = root;
  let databases = new Map<string, Url>();
  Object.keys(result.config.databases).forEach((databaseName) => {
    databases[databaseName] = new Url(result.config.databases[databaseName]);
  });
  return {
    embraceSQLRoot: result.config.embraceSQLRoot,
    databases,
  };
};

/**
 * With a configuration in hand, set up the root context.
 */
export const buildRootContext = async (
  configuration: Configuration
): Promise<RootContext> => {
  return {
    databases: await embraceDatabases(configuration),
  };
};
