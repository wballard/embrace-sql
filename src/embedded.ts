import { InternalContext, buildInternalContext } from "./context";
import { loadConfiguration } from "./configuration";

/**
 * Given a root directory, return an embedded context ready to be used with
 * and inprocess client.
 */
export const buildEmbeddedContext = async (
  root: string
): Promise<InternalContext> => {
  const configuration = await loadConfiguration(root);
  return buildInternalContext(configuration);
};
