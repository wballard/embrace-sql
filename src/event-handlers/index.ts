import { InternalContext, DatabaseInternal } from "../context";
import readFile from "read-file-utf8";
import walk from "ignore-walk";
import path from "path";
import md5 from "md5";
import sqlModulePipeline from "./sqlmodule-pipeline";
import contextPipeline from "./context-pipeline";
import limit from "p-limit";

// thottling
const oneAtATime = limit(1);

/**
 * Scrub up identifiers to be valid JavaScript names.
 */
export const identifier = (key: string): string => {
  const id = key
    .trim()
    // quotes won't do
    .replace(/['"]+/g, "")
    // snake case
    .replace(/\W+/g, "_");
  return /^\d/.test(id) ? "_" + id : id;
};

/**
 * Manage a directory of event handlers, which form the basis of APIs
 *
 * All SQL files in the directory serve as entry points. The process is to enumerate
 * all of these SQL files and build an in memory file system analog with hashes
 * to be used as cache keys.
 *
 * * Loads up and hashes all sql files found
 * * Creates a database specific wrapper for invoking each SQL.
 * * Makes sure all the needed event handlers are generated
 *
 * @param rootContext - like other internal methods, run off the root context
 */
export const embraceEventHandlers = async (
  rootContext: InternalContext
): Promise<InternalContext> => {
  // this should be the only place where a file walk happens
  const fileNames = await walk({
    ignoreFiles: [".sqlmoduleignore"],
    path: rootContext.configuration.embraceSQLRoot,
  });
  // just the SQL files
  const sqlFileNames = await fileNames.filter((fileName) =>
    fileName.toLowerCase().endsWith(".sql")
  );
  // root folders are databases, so attach there
  const allSQLModules = await sqlFileNames.map(async (SQLFileName) => {
    const parsedPath = path.parse(SQLFileName);
    const segments = parsedPath.dir
      .split(path.sep)
      .map((segment) => identifier(segment));
    // database is the first segment
    const databaseName = segments[0];
    // tail end of the path after the database, cleaned up names
    // to be identifiers -- path separator in the URl style
    const pathAfterDatabase = [...segments.slice(1), parsedPath.name]
      .map(identifier)
      .join("/");
    // working with full paths from here on out, one less thing to worry about
    const fullPath = path.join(
      rootContext.configuration.embraceSQLRoot,
      SQLFileName
    );
    const restPath = SQLFileName.replace(/\.sql$/, "");
    // take the path segments and build up a path list
    const handlerPaths = segments.map((_segment, index, array) => {
      return path.join(...array.slice(0, index + 1));
    });
    // get all the 'read' IO done
    const sql = await readFile(fullPath);
    // data about each SQL module
    const sqlModule = {
      restPath,
      fullPath,
      // before handlers run from the root down toward the sql file
      beforeHandlerPaths: handlerPaths,
      // and after handlers are in reverse, from the sql file back toward the root
      afterHandlerPaths: [...handlerPaths].reverse(),
      sql,
      cacheKey: md5(sql),
      contextName: identifier(path.join(parsedPath.dir, parsedPath.name)),
    };
    // collate each module by the containing database
    rootContext.databases[databaseName].SQLModules[
      pathAfterDatabase
    ] = sqlModule;
    return sqlModule;
  });
  // checkpoint -- wait for finish
  await Promise.all(allSQLModules);
  // with all sql modules enumerated time to build up metadata from each database
  const allDatabases = Object.values(rootContext.databases).flatMap(
    async (database: DatabaseInternal) => {
      try {
        // one big transaction around all oof our module building
        // so we can roll back and know we didn't modify our database
        database.transactions.begin();
        // one at a time iteration -- our database isn't re-entrant
        const waitForThem = Object.values(
          database.SQLModules
        ).map(async (sqlModule) =>
          oneAtATime(
            async () =>
              await sqlModulePipeline(rootContext, database, sqlModule)
          )
        );
        await Promise.all(waitForThem);
        return database;
      } finally {
        database.transactions.rollback();
      }
    }
  );
  // let them all finish
  await Promise.all(allDatabases);
  // stitch together the full context and final combined files
  // this is the 'pack' phase
  await contextPipeline(rootContext);
  return rootContext;
};
