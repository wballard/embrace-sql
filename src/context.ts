import { Configuration } from "./configuration";
import { embraceDatabases } from "./database-engines";
import { embraceEventHandlers } from "./event-handlers";
import {
  Database,
  SQLModule,
  SQLColumnMetadata,
  SQLParameters,
  SQLRow,
  SQLModuleDirectExecutors,
} from "./shared-context";
import { AST } from "node-sql-parser";
import { SQLModuleInternal } from "./event-handlers/sqlmodule-pipeline";

/**
 * Keep track of individual migration files with this type.
 */
export type MigrationFile = {
  name: string;
  content: string;
};

/**
 * This is the tree of paths derived from SQL files on disk. This is in
 * a compressed path format, so each key can have / in it.
 */
export type SQLModules = {
  [index: string]: SQLModule;
};

/**
 * A single instance of a database for use internally.
 */
export type DatabaseInternal = Database & {
  /**
   * All modules for this database.
   */
  SQLModules: SQLModules;
  /**
   * Execute the sql module query on this database, and
   * promise some result.
   *
   * @param SQLModule - execute this module, returning results
   * @param parameters - name value pairs are the passed parameters
   */
  execute: (
    sqlModule: SQLModule,
    parameters?: SQLParameters
  ) => Promise<SQLRow[]>;
  /**
   * Analyze the passed module and determine the resultset type(s).
   */
  analyze: (sqlModule: SQLModuleInternal) => Promise<SQLColumnMetadata[]>;
  /**
   * Parse out the SQL.
   */
  parse: (SQLModule) => AST[] | AST;
  /**
   * Do a migration.
   */
  migrate: (migrationFile: MigrationFile) => Promise<void>;
  /**
   * Clean close.
   */
  close: () => Promise<void>;
};

/**
 * All the databases from the internal point of view.
 */
export type AllDatabasesInternal = {
  [index: string]: DatabaseInternal;
};

/**
 * The root context is the context of configuration, databases, and sql modules and is
 * used to drive code generation and runtime execution 'in the engine' of EmbraceSQL
 *
 * The type here is a bit different from the context used in handlers, it has more metadata!
 */
export type InternalContext = SQLModuleDirectExecutors & {
  /**
   * The configuration used to build this context.
   */
  configuration: Configuration;
  /**
   * All configured databases, by name.
   */
  databases: AllDatabasesInternal;
  /**
   * Close all DB connections.
   */
  close: () => Promise<void>;
};

/**
 * With a configuration in hand, set up a new rootContext.
 *
 * This is built to be called -- repeatedly if needed. The idea is you can watch, and
 * rebuild a whole new context as needed -- swapping the root context at runtime to
 * hot-reconfigure the system without worrying about any state leaking.
 *
 * @param configuration - build a root context from this configuration.
 */
export const buildInternalContext = async (
  configuration: Configuration
): Promise<InternalContext> => {
  const databases = await embraceDatabases(configuration);
  const internalContext = {
    configuration,
    databases,
    directQueryExecutors: {},
    close: async (): Promise<void> => {
      const waitForThem = Object.values(databases).map((database) =>
        database.close()
      );
      await Promise.all(waitForThem);
    },
  };
  // need the database first, their connections are used
  // to mine metadata
  await embraceDatabases(configuration);
  await embraceEventHandlers(internalContext);
  return internalContext;
};
