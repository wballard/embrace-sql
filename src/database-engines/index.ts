import embraceSQLite from "./sqlite";
import { AllDatabasesInternal } from "../context";
import { DatabaseInternal } from "../context";
import pLimit from "p-limit";
import {
  SQLModule,
  SQLColumnMetadata,
  SQLRow,
  SQLParameters,
} from "../shared-context";
import { SQLModuleInternal } from "../event-handlers/sqlmodule-pipeline";
import { Configuration } from "../configuration";

/**
 * Serialize database use per process as we really only have one connection.
 */
const oneAtATime = pLimit(1);

/**
 * Embrace a database, bringing it into context.
 *
 * TODO: this needs a wrapper around each database to try to heal itself
 * and reconnect, or to fault and let the container worker process end.
 */
const embraceSingleDatabase = async (
  configuration: Configuration,
  databaseName: string
): Promise<DatabaseInternal> => {
  const dbUrl = configuration?.databases[databaseName];
  switch (dbUrl.protocol.split(":")[0].toLowerCase()) {
    case "sqlite":
      return embraceSQLite(configuration, databaseName);
    default:
      return undefined;
  }
};

/**
 * Every database in the configuraton is embraced and brought into the context.
 */
export const embraceDatabases = async (
  configuration: Configuration
): Promise<AllDatabasesInternal> => {
  // name value pairs inside the root context -- there isn't a strongly
  // typed generated context type available to the generator itself
  const databases = {};
  Object.keys(configuration.databases).forEach(async (databaseName) => {
    const database = await embraceSingleDatabase(configuration, databaseName);
    databases[databaseName] = {
      ...database,
      // here is wrapping the individual database driver execute and analyze
      // with a throttled promise limit -- since we have only one connection
      // these are not re-entrant
      execute: async (
        sqlModule: SQLModule,
        parameters?: SQLParameters
      ): Promise<SQLRow[]> => {
        const results = await oneAtATime(() =>
          database.execute(sqlModule, parameters)
        );
        return results as SQLRow[];
      },
      analyze: async (
        sqlModule: SQLModuleInternal
      ): Promise<SQLColumnMetadata[]> => {
        const results = await oneAtATime(() => database.analyze(sqlModule));
        return results as SQLColumnMetadata[];
      },
    };
  });
  return databases;
};
