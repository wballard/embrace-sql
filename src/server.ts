import { InternalContext } from "./context";
import Koa from "koa";
import bodyparser from "koa-bodyparser";
import OpenAPIBackend from "openapi-backend";
import YAML from "yaml";
import readFile from "read-file-utf8";
import path from "path";
import { HasContextualSQLModuleExecutors } from "./shared-context";

/**
 * Create a HTTP server exposing an OpenAPI style set of endpoints for each Database
 * and SQLModule.
 *
 * This does not actually start the server, it just hands you an instance that you
 * can start listening to later.
 *
 * @param rootContext - root server context with configuration
 * @param executionMap - context name to execution function mapping to actually 'run' a query
 */
export const createServer = async (
  rootContext: InternalContext & HasContextualSQLModuleExecutors
): Promise<Koa<Koa.DefaultState, Koa.DefaultContext>> => {
  const server = new Koa();

  // generated configuration is loaded up
  // TODO: the configuration is left on disk for current debugging -- but probably should be hidden
  const definition = YAML.parse(
    await readFile(
      path.join(rootContext.configuration.embraceSQLRoot, "openapi.yaml")
    )
  );

  const handlers = {};

  // go ahead and make a handler for both GET and POST
  // some of these GET handlers may not be connected at the OpenAPI layer
  // but a few extra functions isn't going to hurt anything
  Object.keys(rootContext.directQueryExecutors).forEach((contextName) => {
    handlers[`get__${contextName}`] = async (
      _openAPI,
      httpContext
    ): Promise<void> => {
      try {
        // parameters from the query
        const context = {
          parameters: httpContext.request.query,
          results: [],
        };
        await rootContext.contextualSQLModuleExecutors[contextName](context);
        httpContext.body = context.results;
        httpContext.status = 200;
      } catch (e) {
        console.error(e);
        httpContext.status = 500;
        httpContext.body = e;
      }
    };
    handlers[`post__${contextName}`] = async (
      _openAPI,
      httpContext
    ): Promise<void> => {
      try {
        // parameters from the body
        const context = {
          parameters: httpContext.request.body,
          results: [],
        };
        await rootContext.contextualSQLModuleExecutors[contextName](context);
        httpContext.body = context.results;
        httpContext.status = 200;
      } catch (e) {
        console.error(e);
        httpContext.status = 500;
        httpContext.body = e;
      }
    };
  });

  // merge up all the handlers just created with the OpenAPI definition
  // and we are ready to go -- this counts on OpenAPI doing some validation
  // before our handlers get called
  const api = new OpenAPIBackend({
    definition,
    handlers,
  });
  api.init();

  // use as koa middleware
  server.use(bodyparser());
  server.use((ctx) => api.handleRequest(ctx.request, ctx));

  // looks like we can get away without any code generation, and just building a
  // map of operation id based handlers
  // https://github.com/anttiviljami/openapi-backend/blob/master/examples/koa/index.js
  return server;
};
