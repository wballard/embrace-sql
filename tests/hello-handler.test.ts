import path from "path";
import fs from "fs-extra";
import { loadConfiguration } from "../src/configuration";
import { buildInternalContext, InternalContext } from "../src/context";
import { createServer } from "../src/server";
import request from "supertest";
import rmfr from "rmfr";
import http from "http";

/**
 * Let's test handlers. These rely on generated code, so the internal context
 * needs to be decorated with the handlers.
 *
 * The setup in beforeAll is a rough simulation of using the system there is a context
 * and directory, with some custom code inserted into handlers.
 */
describe("hello world with a handler", () => {
  let rootContext: InternalContext;
  let listening: http.Server;
  let callback;
  beforeAll(async () => {
    const root = path.relative(process.cwd(), "./.tests/hello-handler");
    // clean up
    await fs.ensureDir(root);
    await rmfr(root);
    // get the configuration and generate - let's do this just the once for speed
    const configuration = await loadConfiguration(root);
    // set up
    await fs.copy(path.join(__dirname, "configs/hello-handler"), root);
    rootContext = await buildInternalContext(configuration);
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { decorateInternalContext } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "context"
    ));
    const server = await createServer(decorateInternalContext(rootContext));
    callback = server.callback();
    listening = server.listen(45679);
    // non logging
    rootContext.configuration.logLevels = [];
  });
  afterAll(async (done) => {
    listening.close(() => done());
  });
  it("will honor handlers over an HTTP call - GET", async () => {
    const response = await request(callback).get(
      "/default/hello?stuff=whirled"
    );
    expect(response.text).toMatchSnapshot();
  });
  it("will honor handlers over an HTTP call - GET - nested", async () => {
    const response = await request(callback).get(
      "/default/nested/echo?stuff=zzzooom"
    );
    expect(response.text).toMatchSnapshot();
  });
  it("will honor handlers over an HTTP call - POST", async () => {
    const postResponse = await request(callback)
      .post("/default/hello")
      .send({ stuff: "amazing" });
    expect(postResponse.text).toMatchSnapshot();
  });
  it("will honor handlers over an HTTP call - POST", async () => {
    const postResponse = await request(callback)
      .post("/default/nested/echo")
      .send({ stuff: "marklar" });
    expect(postResponse.text).toMatchSnapshot();
  });
  it("will honor handlers with a client call", async () => {
    // client
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node"
    ));
    const client = EmbraceSQL("http://localhost:45679");
    expect(
      await client.databases.default.hello.sql({ stuff: "nodey thing" })
    ).toMatchSnapshot();
  });
  it("will honor handlers with an embeddable engine", async () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node-inprocess"
    ));
    const client = EmbraceSQL(rootContext);
    expect(
      await client.databases.default.hello.sql({ stuff: "hole" })
    ).toMatchSnapshot();
  });
  it("will return errors via HTTP - GET", async () => {
    const response = await request(callback).get("/default/hello?stuff=error");
    expect({ status: response.status, text: response.text }).toMatchSnapshot();
  });
  it("will return errors via HTTP - POST", async () => {
    const response = await request(callback)
      .post("/default/hello")
      .send({ stuff: "error" });
    expect({ status: response.status, text: response.text }).toMatchSnapshot();
  });
  it("will throw errors back out to the client", async () => {
    // client
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node"
    ));
    const client = EmbraceSQL("http://localhost:45679");
    // simulated error throw
    expect(
      client.databases.default.hello.sql({ stuff: "error" })
    ).rejects.toMatchSnapshot();
  });
  it("will throw errors back out to the client - nested", async () => {
    // client
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node"
    ));
    const client = EmbraceSQL("http://localhost:45679");
    // simulated error throw
    expect(
      client.databases.default.nested.echo.sql({ stuff: "error" })
    ).rejects.toMatchSnapshot();
  });
});
