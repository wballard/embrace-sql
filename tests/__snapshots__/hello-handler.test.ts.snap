// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello world with a handler will honor handlers over an HTTP call - GET - nested 1`] = `"[{\\"echo\\":\\"zzzooom\\"},{\\"echo\\":\\"zzzooom\\"}]"`;

exports[`hello world with a handler will honor handlers over an HTTP call - GET 1`] = `"[{\\"message\\":\\"whirled!!!\\"},{\\"message\\":\\"whirled!!!\\"},{\\"message\\":\\"whirled!!!\\"},{\\"message\\":\\"whirled!!!\\"}]"`;

exports[`hello world with a handler will honor handlers over an HTTP call - POST 1`] = `"[{\\"message\\":\\"amazing!!!\\"},{\\"message\\":\\"amazing!!!\\"},{\\"message\\":\\"amazing!!!\\"},{\\"message\\":\\"amazing!!!\\"}]"`;

exports[`hello world with a handler will honor handlers over an HTTP call - POST 2`] = `"[{\\"echo\\":\\"marklar\\"},{\\"echo\\":\\"marklar\\"}]"`;

exports[`hello world with a handler will honor handlers with a client call 1`] = `
Array [
  Object {
    "message": "nodey thing!!!",
  },
  Object {
    "message": "nodey thing!!!",
  },
  Object {
    "message": "nodey thing!!!",
  },
  Object {
    "message": "nodey thing!!!",
  },
]
`;

exports[`hello world with a handler will honor handlers with an embeddable engine 1`] = `
Array [
  Object {
    "message": "hole!!!",
  },
  Object {
    "message": "hole!!!",
  },
  Object {
    "message": "hole!!!",
  },
  Object {
    "message": "hole!!!",
  },
]
`;

exports[`hello world with a handler will return errors via HTTP - GET 1`] = `
Object {
  "status": 500,
  "text": "{\\"logLevel\\":\\"error\\",\\"source\\":\\".tests/hello-handler/default/before.ts#8\\",\\"message\\":\\"Error: Simulated Error\\",\\"additional\\":[]}",
}
`;

exports[`hello world with a handler will return errors via HTTP - POST 1`] = `
Object {
  "status": 500,
  "text": "{\\"logLevel\\":\\"error\\",\\"source\\":\\".tests/hello-handler/default/before.ts#8\\",\\"message\\":\\"Error: Simulated Error\\",\\"additional\\":[]}",
}
`;
